### ISSUE TITLE
Pattern synonyms deriving Read

### ISSUE THREAD

**Issue by AntC2** [ghc-proposals/ghc-proposals#142]
Pattern synonyms in expressions are great for providing a less error-prone/shorthand way to construct values that would be complex using the raw constructors from the datatype; or to provide smart constructors that keep the raw constructors abstract.

Then a natural thing to want to do is use those patterns in Strings; and `read` the String to get a value. (Probably the pattern is embedded inside some more complex literal.) Problem: there's no way to derive `Read` instances for patterns. Furthermore instances are for the _type_ not the constructor(s), and in general there's no way for ghc to know all the patterns declared for some type. If the code declares the type `... deriving (Read, ...)`, we can't insert any `read` overloading later for the patterns.

**Hopelessly ideal solution**

    pattern ...                       -- must be bidirectional, either implicit or explicit
      deriving (Read)

(I can't think of any other class than `Read` which it makes sense to derive.)

**Workable approach(?)**

* On the base data type, do not put `... deriving (Read)`.
* Then extend `StandaloneDeriving`:

        deriving instance Read (datatype)              -- standard deriving instance
          pattern (conid1, conid2, ..., conidn)

Downsides:

* If you're importing the base data type decl, and it already has `... deriving (Read)`, you're sunk.
* Or if you want to add some patterns to those already provided with `Read` instances.
* What if you don't want a `Read` instance for the raw data constructors?  
  Do we require mentioning them in the list of _conid_?

If you're avoiding the `... deriving (Read)` on the data type, then yes you could hand-craft `instance Read (datatype) where ...` and overload `read` to parse all the pattern constructors. That's tedious and error-prone code; far better to have the compiler generate it.

**Clumsy/ugly approach**

Provide a callout option/mechanism from `deriving (Read)` for the data type: if the parse fails on the base constructors:

* Parse a _conid_ from the front of `read`'s input string.
* Call method `patternRead` of class `PatternRead`, with two type parameters:
* The data type to be returned (as with `Read`).
* A `Symbol`, being the _conid_ promoted to a type-level String.
* Then we can generate an instance with this pattern decl:

        pattern ...
          deriving (PatternRead)

* Hmm: what about infix _conids_? They might be a long way from the front of `read`'s input string, with tricky parsing needed.

Producing a type-level `Symbol` will need a heavy amount of dependent-typing. And could fail at runtime: no instance. But if we provide _conid_ as merely a String, we can't freely add patterns to `PatternRead`: there could be only one instance per datatype and it would need to 'know' all the _conids_ in advance.

Some other type-level mechanism for chaining together attempts to parse a pattern for the type?

**Comment by mstksg** [comment]
If you're importing a base data type that already has a `Read` instance...do you really actually want to redefine `read` for it?  `Read` as a typeclass is associated with a type, and the `read` function is type-directed.

Since typeclass instances are global, you could have a library that added a pattern synonym for `Int`, and all of a sudden, every library in your entire project now has a different `Read` instance for `Int` ...

**Comment by mstksg** [comment]
The only thing that might make sense to me is to have allow a data type's **canonical** `Read` and `Show` instances be derived and provided in terms of a pattern synonym (somewhat in a *DerivingVia*-style API), or maybe multiple.  Once declared, it can't be added to or modified by any other module or declaration.

**Comment by AntC2** [comment]
> do you really actually want to redefine `read` for it? 

No I didn't say "redefine". I'm assuming the data constructors introduced for the pattern are distinct from those in the `data` decl. Then `read` should be able to parse _either_/any constructors. Given

    data Type           = App String  [Type]

    pattern Arrow t1 t2 = App "->"    [t1, t2]
    pattern Int         = App "Int"   []
    pattern Maybe t     = App "Maybe" [t]

These should produce the same result:

    x1 = read "App \"->\" [App \"Int\" [], App \"Int\" []]" :: Type  -- data type's constructors
    x2 = read "Arrow Int Int" :: Type                                -- pattern's constructors
    x3 = read "Arrow Int (App \"Int\" [])" :: Type                   -- mixing constructors

Which literal is a user more likely to get right?

Then ideally (this is my "hopeless" solution); any module can import the `data` decl; add some patterns; and make those patterns available for `read` _in addition to_ those in the `data` decl.

**Comment by mstksg** [comment]
Yes, I do agree that if it happens, the `Read` derivation or definition should be all in one location, and not separated out at each pattern site.

> Then ideally (this is my "hopeless" solution); any module can import the data decl; add some patterns; and make those patterns available for read in addition to those in the data decl.

Do you mean only in the case where that declaration does not have a `Read` instance already?  (So, not including the situation of someone exporting your example that derives `Read` already based)
